Feature/스터디-생성-수정-validation

mieumje opened the pull request:
<!-- 
  PR 작성 시 해당 Issue 연결
  PR 이름: Feature/개발-완료-기능
  PR은 선착순 2명이 달면 수정해서 merge 
-->

## ⚙기능
<!-- 기능을 설명해주세요 -->
우선 커밋 단위가 크고, style과 섞여 있는 점 죄송합니다😓 다음부턴 조금 더 깔끔하게 분리 작성 할 수 있도록 하겠습니다.
스터디 생성과 수정이 공통으로 수정된 부분이 있어 Edit Form도 같이 수정된 부분이 있습니다.
![ezgif com-gif-maker](https://user-images.githubusercontent.com/66072832/183007724-2e0dc4b6-b9dc-45c7-9b5b-2503bc4d0288.gif)

### 제목, 분야, 인원
![image](https://user-images.githubusercontent.com/66072832/183007846-5acabfcd-f24f-4e8a-a763-40950a52bf9a.png)

### 상세 설명
![image](https://user-images.githubusercontent.com/66072832/183007890-83dd6c7c-f8a1-436e-a975-0997bcaac1c8.png)

title과 description은 formik, yup을 통해 validation 했습니다. 분야, 인원, 이미지 등은 커스텀하여 검증하도록 했습니다.
Form을 제출하게 되면 required 필드 검증을 하고, 값이 비어있거나 잘못 되었다면 에러 메세지를 노출하도록 설정되어 있습니다.

Form 검증이 끝나고 제출이 되면, setTimeout 함수로 API 요청 로딩 처리를 임시로 구현해 두었고, 요청을 보내는 동안 버튼이 disabled됩니다. 이후 전송할 Form Data를 console.log()로 출력하도록 되어 있습니다.

## 📃구현 내용
<!-- 구현 내용을 작성해 주세요 -->
- 스터디 생성 페이지 Form validation
  - title : 최소 5, 최대 50자 제한 formik validation / trim() strict()를 통해 앞, 뒤 공백을 제한 / required 필드
  - description : 최소 10, 최대 1,000자 제한 formik validation / trim()을 통해 공백 제외 글자 제한 / submit 시 trim()한 value를 전달 / required 필드
  - 그 외 필드
    - 분야 : 제출 버튼 onClick 시 검증 / required 필드
    - 인원 : 제출 버튼 onClick 시 검증 / required 필드
    - 이미지 업로드 : 파일 업로드 시 검증(파일 크기, 파일 타입) / 업로드 로딩 시 Spinner 처리

Select 컴포넌트를 어제 키아님이 만들어준 공통 Select 컴포넌트로 대체했고, 옵션도 키아님이 만들어 준 곳에 스티더 인원 옵션 추가해서 사용했습니다. cb96ec8834bd3bf209a515ba92d3f080c0cf06b6 83f59551f8a65f8cde2edc2f2f42bf8b26b64ad5  9e3786bc9c9fd855665e2f4190a89d825723f24e 163f0fc1d169b1fa12e61f5ea31ca08130caab0d

API 요청 시 보내기 위한 FormData 함수를 분리해 작성했습니다. 4c0aa465c6f732d62eda3a8dc0e7e3439697e2d8

이미지 업로드 로딩 시 Spinner 처리 하도록 했습니다. d457924a8c16ff8b1f5a1947bf3a3df23463102e

전역 객체에서 사용자 MBTI를 받게 되면 해당 사용자의 MBTI에 맞는 추천 MBTI가 담긴 배열을 반환하는 함수를 추가했습니다. a00371d3ff08fb55014f1543fb3ee6e28710bb8e

팽님이 의견 주셨던 checkbox ripple disabled 추가했습니다. 163f0fc1d169b1fa12e61f5ea31ca08130caab0d

### 8월 5일 추가 내용
1. submit 시 title, description만 validation을 통과하면 나머지 error를 체크하지 않는 문제 -> checkError 함수 추가 7f9f1188769aa9eb6c076e05fd976fcf7c0d56f4, e2aa4aa006a7be77a97b232a795a6f544d84298a
2. 리뷰 반영 내용 -> a8b4354c65bf1d394f1987b831d3cae9585d34d2, d3205b4391101028422242929889459a038d1235, d4cbfa32284bc02ee9cc63780bf7d3004d464c05
3. 스크럼 회의 후 Select, Input 컴포넌트 error message를 커스텀하지 않고 helperText, error를 사용하기로 한 내용 반영 -> 2e918c57e0c5d096b6e694020cb41c04debe5ac7, b57924767fe6294af84fd861783baef1374d793b
![image](https://user-images.githubusercontent.com/66072832/183084390-d1d020a9-23ee-4058-b1c3-1a6fd0d2c06c.png)
![image](https://user-images.githubusercontent.com/66072832/183084433-008f5ed2-4c65-4958-b759-c16ef979a04f.png)


## 💡기타
<!-- 추가로 적고 싶은 내용을 모두 적어주세요 -->
formik, yup과 커스텀 validation을 같이 써서 문제가 생기지 않을까 하는 우려가 있습니다.. validation 자체가 되지 않으면 submit이 되지 않아서 일단은 괜찮아 보이는데 엣지 케이스를 많이 생각해보지 못한 것 같습니다.

mieumje committed:
refactor: Select 컴포넌트 공통 컴포넌트로 변경

mieumje committed:
refactor: 스터디 생성 Form styled 컴포넌트 수정 및 추가

InputWrapper, TopicWrapper, CameraIcon,  ErrorMessage 컴포넌트 추가

LocationWrapper, PeopleWrapper, FileUploadWrapper
ImageContainer, ImageWrapper, Image, ButtonWrapper 수정

mieumje committed:
feat, refactor: form validation 추가, state, 함수 정리

title, description validation 추가
그 외 필드(topic, 지역, location, image)는 커스텀

title, body local state 삭제
errorMessage local state 추가

onImageChange 함수 수정
errorMessage handle useEffect 추가
topic, region, recruitsNumber state handle 함수 추가 및 적용
FormData 생성 함수 추가

mieumje committed:
feat: selectOptions에 recruitsNumberOptions 추가

mieumje committed:
refactor: Select options 변경

mieumje committed:
refactor: mbtiOptions로 수정, disableRipple 추가

mieumje committed:
refactor: Yup Schema 수정, 날짜 toString() 제거

mieumje committed:
refactor: 사용하지 않는 변수 삭제

mieumje committed:
refactor: state type 추가, isOnlineChange 함수 수정

mieumje committed:
refactor: Input 컴포넌트 교체

mieumje committed:
refactor: Form validation trim 추가

title의 경우 앞, 뒤 공백 모두 제한
description의 경우 공백 제외 길이만 제한, formdata에서 description 값을 trim()

mieumje committed:
feat: 이미지 업로드 중 Spinner 처리

mieumje committed:
style: description input height 500px로 지정

mieumje committed:
feat: recommendMbtis 함수 생성

사용자 mbti를 기준으로 추천 mbti array를 return

GeonWooPaeng commented on src/containers/StudyCreateForm/StudyCreateForm.tsx:
```diff
@@ -167,9 +99,49 @@ function StudyCreateFormContainer() {
     }
   };
 
+  const recommendMbtis = useCallback((mbti: string) => {
```
p4;

swich문을 사용하는 것도 좋아보이지만
다음과 같이 객체로 관리하는 것은 어떨까요?

```ts
{
  'ENTJ': ['INFP', 'ISFP']
}
```

mieumje replied:
이 방법이 더 괜찮은 것 같네요! 감사합니다👍24fdca07c329daf9ddb189325e87bc060e8c1ac8

GeonWooPaeng commented on src/containers/StudyCreateForm/StudyCreateForm.tsx:
```diff
@@ -64,97 +53,40 @@ const radioValues = [
   const [imageSrc, setImageSrc] = useState<File | null>(null);
   const [thumbnailImage, setThumbnailImage] = useState<string>('');
-  const [body, setBody] = useState('');
+  const [fileErrorMessage, setFileErrorMessage] = useState<string>('');
+  const [topicErrorMessage, setTopicErrorMessage] = useState<string>('');
+  const [peopleErrorMessage, setPeopleErrorMessage] = useState<string>('');
```
p4;
Error Message를 관리하는 State를 1개로 줄여보는 것은 어떨까요?
```suggestion
  const [ErrorMessage, setErrorMessage] = useState<object>({file: '', topic:'', people: ''});
```

mieumje replied:
제가 코드를 잘 짰으면 괜찮았을 수 있는데, error message를 하나의 객체 state로 관리하면
```js 
const checkSelectOptions = () => {
    if (!topic) {
      setTopicErrorMessage('분야를 선택해주세요');
      setErrorMessages({ ...errorMessages, topic: '분야를 선택해주세요'});
    }
    if (!recruitsNumber) {
      setPeopleErrorMessage('인원수를 선택해주세요');
      setErrorMessages({ ...errorMessages, people: '인원수를 선택해주세요'});
    }
}
```
이 함수에서 topic에 대한 state가 변경이 되지 않습니다ㅜㅜ.. 저도 정확히는 모르는데 useState의 setState 반영이 우리가 생각하는 만큼 바로 반영되지 않는 것 같더라구요... 이부분에 대해서는 조금 더 알아볼 필요가 있습니다.. 이러한 이유로 우선 변경 없이 남겨두었습니다..! 코멘트 감사합니다 팽님👍

GeonWooPaeng commented on src/containers/StudyCreateForm/StudyCreateForm.tsx:
```diff
@@ -178,166 +150,296 @@ function StudyCreateFormContainer() {
-    console.log('선호 MBTI ' + mbtiCheckedList);
-    console.log('이미지 ' + imageSrc);
-    console.log('내용' + body);
+  useEffect(() => {
+    if (topic) setTopicErrorMessage('');
+  }, [topic]);
+
+  useEffect(() => {
+    if (recruitsNumber) setPeopleErrorMessage('');
+  }, [recruitsNumber]);
```
p4;
2개로 나눈 이유가 궁금합니다. :) 
저도 React가 약해서 ㅜㅜ

```suggestion
  useEffect(() => {
    if (topic) setTopicErrorMessage('');
    if (recruitsNumber) setPeopleErrorMessage('');
  }, [topic, recruitsNumber]);
```

jaeung-E replied:
의존성 배열에 들어가는 내용이 달라서 그러신 것 같습니다! 😁

팽님이 서제스트 해주신 내용대로라면, `recruitsNumber, topic` 중 하나의 값만 바뀌어도 두 에러메시지가 모두 초기화됩니다.

mieumje replied:
맥키님이 말씀해주신 대로 의도했습니다! `topic`과 `recruitsNumber` 에러 메세지를 초기화 하려고 분리했는데 더 나은 방법은 생각해내지 못해서..😓

GeonWooPaeng replied:
아하 감사합니다. :)

GeonWooPaeng commented on src/containers/StudyCreateForm/StudyCreateForm.tsx:
```diff
@@ -64,97 +53,40 @@ const radioValues = [
-    label: '10',
-  },
-];
+const CreateSchema = Yup.object({
+  title: Yup.string()
+    .trim('앞, 뒤 공백을 제거해주세요.')
+    .strict()
+    .max(50, '50자를 넘을 수 없습니다.')
+    .min(5, '최소 5자 이상입니다.')
+    .required('제목을 입력해주세요.'),
+  description: Yup.string()
+    .trim()
+    .max(1000, '1,000자를 넘을 수 없습니다.')
+    .min(10, '최소 10자 이상입니다.')
+    .required('스터디 내용을 입력해주세요.'),
+});
```
오 이렇게 빼서 다른 파일로 관리할 수 있겠네요 :)

mieumje replied:
스터디 생성이랑, 수정에서 Yup 객체가 동일하게 사용되는데 어디에서 이 녀석을 관리해야 할 지 잘 모르겠네요..

GeonWooPaeng replied:
맞네요 이건 refactoring 할때 같이 생각해보죠 :)

GeonWooPaeng commented on src/containers/StudyCreateForm/StudyCreateForm.tsx:
```diff
@@ -178,166 +150,296 @@ function StudyCreateFormContainer() {
+        }, 3000);
+      }}
+    >
+      {(formik) => {
```
p4;
`formik.handleSubmit, formik.isSubmitting` 같은 것을 빼서 사용해도 좋을 것 같습니다.
```suggestion
      {({handleSubmit, isSubmitting}) => {
```

mieumje replied:
감사합니다👍 반영 완료했습니다! 8a85cf0cd105803b237ac0928be7d668d41ff387

GeonWooPaeng reviewed (approved):
Jay 정말 수고하셨습니다. !! Loading도 넣어주시고 많은 error처리를 해주셨네요!! 
기능적인 문제는 크게 없고 같이 이야기한 Error Message 위치만 맞추면 될 것 같습니다.
코드를 줄이는 방법만 Comment 한 점 죄송합니다. :) 
편하게 보시고 가볍게 넘어가셔도 됩니다.

jaeung-E commented on src/components/StudyCreate/StudyCreateText/StudyCreateMbtiRecommend.tsx:
```diff
@@ -73,23 +73,26 @@ const setMbtiColor = (mbti: string) => {
-      입니다.
-    </TextWrapper>
-  );
+  if (mbtis?.length) {
```
만약 `User` 객체의 `mbti`가 존재하지 않을 경우를 대비해서 분기처리를 하신걸로 이해했습니다!

회원가입 시 `MBTI`를 필수적으로 선택하도록 되어있기 때문에, `mbti prop`에는 항상 특정 값이 넘어 올 예정이므로 해당 변경사항은 이전으로 되돌리셔도 문제 없을 것 같습니다!

mieumje replied:
아하 감사합니다! mbti 선택이 필수였군요 ㅎㅎ 66a8b368fbfbfa0efede48d90aa2ae9545900e1c

jaeung-E reviewed (commented):
`Validation` 구현하느라 고생 많으셨습니다 제이님!!😆 다만 지금은 몇 가지 이슈가 있는 것 같습니다!!

직접 `suggestion`으로 수정을 도와드리고 싶지만, `formik, yup`을 적용해보지 않아서 디테일한 리뷰는 못 드려서 죄송합니다.😥

일단 제가 찾은 이슈는 아래와 같습니다!

- `title, contents`의 `valiadation` 검사 후 유효하다면 그 외 다른 필드 값의 유무와 상관없이 `form`이 제출됨.

해당 이슈로 인해서 `title, contents`의 길이 유효성만 통과한다면 분야, 인원을 선택하지 않았을 경우 에러 메시지는 출력되나, 해당 값이 비어져 있는 상태로 정상적으로 폼이 제출됩니다!

mieumje committed:
feat: checkError 함수 추가

title, description 외의 validation error가 있는 경우 true return 한다.

함수가 true를 return하는 경우, onSubmit을 중단한다.

mieumje committed:
chore: Mbti 추천 컴포넌트 다시 원상태로 복구

user 객체에 mbti 프로퍼티는 필수라서 mbti가 비어있는지 확인하지 않는다.

mieumje committed:
refactor: mbti 추천 함수를 상수 객체로 변경

mieumje committed:
refactor: formik에 쓰이는 요소 구조 분해 할당

mieumje commented:
> * `title, contents`의 `valiadation` 검사 후 유효하다면 그 외 다른 필드 값의 유무와 상관없이 `form`이 제출됨.
> 
> 해당 이슈로 인해서 `title, contents`의 길이 유효성만 통과한다면 분야, 인원을 선택하지 않았을 경우 에러 메시지는 출력되나, 해당 값이 비어져 있는 상태로 정상적으로 폼이 제출됩니다!

@jaeung-E  남아있는 error가 있는지 확인하는 함수를 추가했습니다.  dc470b1f600f878d948a1f9f42d51fa973e78921

mieumje committed:
refactor: helperText, error 적용 / region 에러 메세지 추가

Select 필드 에러메세지를 TextField의 helperText로 대체
isOnline flag가 false고 region state가 없는 경우 region error message 추가
region error message handler 추가

mieumje committed:
refactor: Input helperText 적용 / checkError 수정

LabelInput, MultiLineInput helperText, error 적용
checkError 함수에 오프라인일 때 region error message가 있는지 체크 추가

mieumje committed:
refactor: 사용하지 않는 코드 삭제

mieumje committed:
refactor: checkError 함수 isError 변수 추가

jaeung-E reviewed (approved):
변경사항 확인 후 Approve 했습니다! 

확실히 `helperText` 사용하니 에러메시지 디테일이 좀 더 살아난 것 같습니다 ㅎㅎ 고생 많으셨습니다 제이!!👍
